File: emoji-kitchen/emoji-kitchen.js
const fs = require('fs');
const http = require('http');
const https = require('https');
const path = require('path');

const { execSync } = require("child_process");
const {lookup} = require('lookup-dns-cache');

const PORT = 36363;

// https://nodejs.org/api/http.html#http_http_get_options_callback
const REQUEST_OPTS = {
  family: 4,
  lookup: lookup,
};

String.prototype.title = function() {
  return this.replace(/(^|\s)\S/g, (t) => t.toUpperCase());
};

// Negative indices for arrays
Array.prototype.get = function(i) {
  return this[(i + this.length) % this.length];
};

Array.prototype.makeMap = function(keyfunc) {
  return this.reduce(
    (dict, item) => {
      dict[keyfunc(item)] = item;
      return dict;
    },
    {});
};

const WF_DIR = process.cwd().split('/emoji-kitchen')[0];
const MAIN_EMOJI_JSON = `${WF_DIR}/alfreditems.json`;

const KITCHEN_DATA_DIR = (() => {
  var wf_data = process.env.alfred_workflow_data;
  if (!wf_data) {
    let HOME = process.env.HOME;
    wf_data = `${HOME}/Library/Application Support/Alfred` +
      '/Workflow Data/mr.pennyworth.fastEmoji';
  }

  let kitchen_data = `${wf_data}/kitchen_data`;
  if (!fs.existsSync(kitchen_data)) {
    fs.mkdirSync(kitchen_data, { recursive: true });
  }
  return kitchen_data;
})();


const emojiToInfoMap =
  JSON
    .parse(fs.readFileSync(MAIN_EMOJI_JSON))
    .items
    .makeMap((i) => stripVarSel(i.variables.emoji));


// All the results of querying emoji-kitchen go to
// the emoji fridge for easy later access.
function updateFridge(alfredItems) {
  let fridgePath = `${KITCHEN_DATA_DIR}/fridge.json`;
  var fridge = {'items': []};
  if (fs.existsSync(fridgePath)) {
    fridge = JSON.parse(fs.readFileSync(fridgePath));
  }
  let existing = new Set(fridge.items.map(i => i.uid));
  alfredItems.forEach((i) => {
    if (!existing.has(i.uid)) {
      fridge.items.push(i);
    }
  });
  fs.writeFileSync(fridgePath, JSON.stringify(fridge));
}

function parseTenorDataAndRespondToAlfred(tenorData, res, emoji, emoji2) {
  let downloadMonitor = new Map();
  let alfredItems = tenorData.results.map((tenorEntry) => {
    let pngPath = download(tenorEntry.url, downloadMonitor);
    let emojis = tenorEntry.tags;
    let emojiInfos = emojis.map((e) => emojiToInfoMap[stripVarSel(e)]);
    let keywords = emojiInfos.map(i => i.match).join(' ').split(' ');

    var uniqueKewords = [...new Set(keywords)].join(' ');
    var title = emojiInfos.map(i => i.title).join(' + ');
    if (emojis.length == 1) {
      uniqueKewords = 'double ' + uniqueKewords;
      title = 'double ' + title;
    }

    return {
      'arg': pngPath,
      'type': 'file:skipcheck',
      'uid': pngPath,
      'title': title,
      'match': uniqueKewords,
      'subtitle': uniqueKewords,
      'icon': {
        'path': pngPath
      }
    };
  });

  var responseItems = alfredItems;
  if (!emoji2) {
    let freshCookItem = {
      'arg': 'cook_new',
      'title': 'Cook Something New!',
      'subtitle': `Add another emoji to ${emoji} to make a new sticker`,
      'icon': {
        'path': 'cook-new.png'
      },
      'variables': {
        'emoji2': emoji
      }
    };
    responseItems = [freshCookItem].concat(alfredItems);
  }
  if (alfredItems.length == 0) {
    let sorryItem = {
      'valid': false,
      'title': 'Sorry, Empty Kitchen!',
      'icon': {
        'path': 'empty-kitchen.png'
      }
    };
    responseItems = [sorryItem].concat(responseItems);
  }

  // Busy-wait for downloads to finish
  var timeout = setInterval(function() {
    if (downloadMonitor.size == tenorData.results.length) {
      res.setHeader('Content-Type', 'application/json');
      res.write(JSON.stringify({'items': responseItems}));
      res.end();
      console.log('Responded to alfred');
      clearInterval(timeout); 
    }
  }, 50);

  updateFridge(alfredItems);
}


function download(pngUrl, downloadMonitor) {
  let pngName = pngUrl.split('/').get(-1);
  let pngPath = `${KITCHEN_DATA_DIR}/${pngName}`;

  if (!fs.existsSync(pngPath)) {
    const stream = fs.createWriteStream(pngPath);
    const getter = pngUrl.startsWith('https:') ? https : http;
    getter.get(pngUrl, REQUEST_OPTS, (response) => {
      console.log(`Downloading ${pngUrl}`);
      response.pipe(stream);
      downloadMonitor.set(pngUrl, true);
    });
  } else {
    downloadMonitor.set(pngUrl, true);
  }

  return pngPath;
}


function sendErrorToAlfred(res, error) {
  res.setHeader('Content-Type', 'application/json');
  res.write(JSON.stringify({
    'items': [{
      'title': 'Error: ' + error.message
    }]
  }));
  res.end();
}

// strip variation selectors
// Example: https://emojipedia.org/variation-selector-16/
function stripVarSel(emoj) {
  let varSel = /([\u180B-\u180D\uFE00-\uFE0F]|\uDB40[\uDD00-\uDDEF])/g;
  return emoj.replace(varSel, '');
}

http.createServer(function (req, res) {
  let url = new URL(req.url, `http://${req.headers.host}`);
  var query = stripVarSel(url.searchParams.get('query'));
  var query2 = stripVarSel(url.searchParams.get('query2'));
  if (query2) {
    [query, query2] = [query2, query];
  }

  let emojiInfo = emojiToInfoMap[query];

  let scriptIconPath = `${WF_DIR}/DBEA5CCE-9222-4700-9C4D-E28F2C222532.png`;
  let script2IconPath = `${WF_DIR}/5D97EA1D-9B07-4355-9BAA-E2C508EEEB02.png`;
  let emojiIconPath = `${WF_DIR}/${emojiInfo.icon.path.replace('./', '')}`;
  execSync(`/bin/cp "${emojiIconPath}" "${scriptIconPath}"`);
  execSync(`/bin/cp "${emojiIconPath}" "${script2IconPath}"`);

  let api = new URL('https://tenor.googleapis.com/v2/featured');
  api.searchParams.append('key', 'AIzaSyAyimkuYQYF_FXVALexPuGQctUWRURdCYQ');
  api.searchParams.append('client_key', 'gboard');
  api.searchParams.append('contentfilter', 'high');
  api.searchParams.append('media_filter', 'png_transparent');
  api.searchParams.append('component', 'proactive');
  api.searchParams.append('collection', 'emoji_kitchen_v5');
  api.searchParams.append('locale', 'en_US');
  api.searchParams.append('country', 'US');
  api.searchParams.append('q', query2 ? `${query}_${query2}` : query);
  console.log(api);
  https.get(api, REQUEST_OPTS, (tenorRes) => {
    const { statusCode } = tenorRes;
    const contentType = tenorRes.headers['content-type'];

    let error;
    // Any 2xx status code signals a successful response but
    // here we're only checking for 200.
    if (statusCode !== 200) {
      error = new Error(
        `Request Failed.\nStatus Code: ${statusCode}`
      );
    } else if (!/^application\/json/.test(contentType)) {
      error = new Error(
        'Invalid content-type.\n' +
          `Expected application/json but received ${contentType}`
      );
    }

    if (error) {
      console.error(error.message);

      // Consume response data to free up memory
      tenorRes.resume();
      sendErrorToAlfred(res, error);
      return;
    }

    tenorRes.setEncoding('utf8');
    let rawData = '';
    tenorRes.on('data', (chunk) => { rawData += chunk; });
    tenorRes.on('end', () => {
      try {
        const tenorData = JSON.parse(rawData);
        parseTenorDataAndRespondToAlfred(tenorData, res, query, query2);
      } catch (e) {
        console.error(e.message);
        res.write(e.message);
        res.end();
      }
    });
  }).on('error', (e) => {
    console.error(`Got error: ${e.message}`);
    sendErrorToAlfred(res, e);
  });

  console.log(url.searchParams);
}).listen(PORT);

File: emoji-kitchen/emoji-kitchen.swift
#!/usr/bin/env swift

import Foundation

let apiKey = "AIzaSyAyimkuYQYF_FXVALexPuGQctUWRURdCYQ"

struct EmojiInfo {
  let title: String
  let match: String
  let iconPath: String
}

enum KitchenError: Error {
  case message(String)
}

func stripVariationSelectors(_ value: String) -> String {
  var scalars = String.UnicodeScalarView()
  for scalar in value.unicodeScalars {
    let codePoint = scalar.value
    let isMongolianVariation =
      codePoint >= 0x180B && codePoint <= 0x180D
    let isVariationSelector =
      codePoint >= 0xFE00 && codePoint <= 0xFE0F
    let isSupplementaryVariation =
      codePoint >= 0xE0100 && codePoint <= 0xE01EF
    if isMongolianVariation ||
      isVariationSelector ||
      isSupplementaryVariation {
      continue
    }
    scalars.append(scalar)
  }
  return String(scalars)
}

func fetchData(from url: URL) throws -> Data {
  let semaphore = DispatchSemaphore(value: 0)
  var dataOut: Data?
  var statusCode = 0
  var contentType = ""
  var fetchError: Error?

  let task = URLSession.shared.dataTask(with: url) {
    data,
    response,
    error in
    dataOut = data
    fetchError = error
    if let httpRes = response as? HTTPURLResponse {
      statusCode = httpRes.statusCode
      contentType = httpRes.value(
        forHTTPHeaderField: "content-type"
      ) ?? ""
    }
    semaphore.signal()
  }
  task.resume()
  semaphore.wait()

  if let fetchError {
    throw fetchError
  }
  if statusCode != 200 {
    throw KitchenError.message(
      "Request Failed. Status Code: \(statusCode)"
    )
  }
  if !contentType.lowercased().contains("application/json") &&
    !url.pathExtension.lowercased().contains("png") {
    throw KitchenError.message(
      "Invalid content-type. Received \(contentType)"
    )
  }
  guard let dataOut else {
    throw KitchenError.message("No response data.")
  }
  return dataOut
}

func ensureDirectory(_ path: String) throws {
  try FileManager.default.createDirectory(
    atPath: path,
    withIntermediateDirectories: true
  )
}

func uniqueKeywords(_ keywords: [String]) -> String {
  var seen = Set<String>()
  var ordered = [String]()
  for keyword in keywords where !keyword.isEmpty {
    if !seen.contains(keyword) {
      seen.insert(keyword)
      ordered.append(keyword)
    }
  }
  return ordered.joined(separator: " ")
}

func loadEmojiMap(from path: String) throws -> [String: EmojiInfo] {
  let data = try Data(contentsOf: URL(fileURLWithPath: path))
  guard
    let root = try JSONSerialization.jsonObject(with: data)
      as? [String: Any],
    let items = root["items"] as? [[String: Any]]
  else {
    throw KitchenError.message("Unable to load alfreditems.json.")
  }

  var map = [String: EmojiInfo]()
  for item in items {
    guard
      let title = item["title"] as? String,
      let match = item["match"] as? String,
      let icon = item["icon"] as? [String: Any],
      let iconPath = icon["path"] as? String,
      let vars = item["variables"] as? [String: Any],
      let emoji = vars["emoji"] as? String
    else {
      continue
    }
    map[stripVariationSelectors(emoji)] = EmojiInfo(
      title: title,
      match: match,
      iconPath: iconPath
    )
  }
  return map
}

func copyIconFiles(workflowDir: String, emojiInfo: EmojiInfo) {
  let fm = FileManager.default
  let iconRel = emojiInfo.iconPath.replacingOccurrences(of: "./", with: "")
  let src = "\(workflowDir)/\(iconRel)"
  let targets = [
    "\(workflowDir)/DBEA5CCE-9222-4700-9C4D-E28F2C222532.png",
    "\(workflowDir)/5D97EA1D-9B07-4355-9BAA-E2C508EEEB02.png"
  ]
  for target in targets {
    do {
      if fm.fileExists(atPath: target) {
        try fm.removeItem(atPath: target)
      }
      try fm.copyItem(
        atPath: src,
        toPath: target
      )
    } catch {
      continue
    }
  }
}

func buildAPIURL(query: String) -> URL? {
  var comps = URLComponents(
    string: "https://tenor.googleapis.com/v2/featured"
  )
  comps?.queryItems = [
    URLQueryItem(name: "key", value: apiKey),
    URLQueryItem(name: "client_key", value: "gboard"),
    URLQueryItem(name: "contentfilter", value: "high"),
    URLQueryItem(name: "media_filter", value: "png_transparent"),
    URLQueryItem(name: "component", value: "proactive"),
    URLQueryItem(name: "collection", value: "emoji_kitchen_v5"),
    URLQueryItem(name: "locale", value: "en_US"),
    URLQueryItem(name: "country", value: "US"),
    URLQueryItem(name: "q", value: query)
  ]
  return comps?.url
}

func downloadPng(
  pngURL: String,
  kitchenDataDir: String,
  workflowDir: String
) throws -> String {
  guard let url = URL(string: pngURL) else {
    throw KitchenError.message("Invalid png URL.")
  }
  let fileName = url.lastPathComponent
  let outPath = "\(kitchenDataDir)/\(fileName)"
  if FileManager.default.fileExists(atPath: outPath) {
    return outPath
  }

  let data = try fetchData(from: url)
  try data.write(
    to: URL(fileURLWithPath: outPath),
    options: .atomic
  )

  let proc = Process()
  proc.executableURL = URL(
    fileURLWithPath: "\(workflowDir)/set_icon.sh"
  )
  proc.arguments = [outPath, outPath]
  try? proc.run()
  proc.waitUntilExit()
  return outPath
}

func updateFridge(
  kitchenDataDir: String,
  newItems: [[String: Any]]
) throws {
  let fridgePath = "\(kitchenDataDir)/fridge.json"
  var fridge: [String: Any] = ["items": []]
  if FileManager.default.fileExists(atPath: fridgePath) {
    let data = try Data(contentsOf: URL(fileURLWithPath: fridgePath))
    if let parsed = try JSONSerialization.jsonObject(with: data)
      as? [String: Any] {
      fridge = parsed
    }
  }

  var items = (fridge["items"] as? [[String: Any]]) ?? []
  var existing = Set<String>()
  for item in items {
    if let uid = item["uid"] as? String {
      existing.insert(uid)
    }
  }
  for item in newItems {
    guard let uid = item["uid"] as? String else {
      continue
    }
    if !existing.contains(uid) {
      existing.insert(uid)
      items.append(item)
    }
  }
  fridge["items"] = items
  let outData = try JSONSerialization.data(withJSONObject: fridge)
  try outData.write(
    to: URL(fileURLWithPath: fridgePath),
    options: .atomic
  )
}

func writeTmpJson(
  kitchenDataDir: String,
  items: [[String: Any]]
) throws -> String {
  let tmpPath = "\(kitchenDataDir)/tmp.json"
  let outObj: [String: Any] = ["items": items]
  let outData = try JSONSerialization.data(withJSONObject: outObj)
  try outData.write(
    to: URL(fileURLWithPath: tmpPath),
    options: .atomic
  )
  return tmpPath
}

func runKitchen() throws -> String {
  if CommandLine.arguments.count != 4 {
    throw KitchenError.message(
      "Usage: emoji-kitchen.bin <emoji1> <emoji2> <kitchen-data-dir>"
    )
  }

  var query = stripVariationSelectors(CommandLine.arguments[1])
  var query2 = stripVariationSelectors(CommandLine.arguments[2])
  let kitchenDataDir = CommandLine.arguments[3]
  let workflowDir = FileManager.default.currentDirectoryPath
  let mainJsonPath = "\(workflowDir)/alfreditems.json"
  try ensureDirectory(kitchenDataDir)
  let emojiMap = try loadEmojiMap(from: mainJsonPath)

  if !query2.isEmpty {
    let q = query
    query = query2
    query2 = q
  }

  guard let emojiInfo = emojiMap[query] else {
    throw KitchenError.message("Unknown emoji query: \(query)")
  }
  copyIconFiles(
    workflowDir: workflowDir,
    emojiInfo: emojiInfo
  )

  let searchQuery = query2.isEmpty ? query : "\(query)_\(query2)"
  guard let apiURL = buildAPIURL(query: searchQuery) else {
    throw KitchenError.message("Failed to build API URL.")
  }
  let data = try fetchData(from: apiURL)
  guard
    let root = try JSONSerialization.jsonObject(with: data)
      as? [String: Any],
    let results = root["results"] as? [[String: Any]]
  else {
    throw KitchenError.message("Invalid Tenor response JSON.")
  }

  var alfredItems = [[String: Any]]()
  for result in results {
    guard
      let pngURL = result["url"] as? String,
      let tags = result["tags"] as? [String]
    else {
      continue
    }
    let pngPath = try downloadPng(
      pngURL: pngURL,
      kitchenDataDir: kitchenDataDir,
      workflowDir: workflowDir
    )
    var infos = [EmojiInfo]()
    for tag in tags {
      let key = stripVariationSelectors(tag)
      if let info = emojiMap[key] {
        infos.append(info)
      }
    }

    let allKeywords = infos
      .map { $0.match }
      .joined(separator: " ")
      .split(separator: " ")
      .map { String($0) }
    var keywords = uniqueKeywords(allKeywords)
    var title = infos.map { $0.title }.joined(separator: " + ")
    if tags.count == 1 {
      keywords = "double \(keywords)"
      title = "double \(title)"
    }
    let item: [String: Any] = [
      "arg": pngPath,
      "type": "file:skipcheck",
      "uid": pngPath,
      "title": title,
      "match": keywords,
      "subtitle": keywords,
      "icon": [
        "path": pngPath
      ]
    ]
    alfredItems.append(item)
  }

  var responseItems = alfredItems
  if query2.isEmpty {
    let freshCook: [String: Any] = [
      "arg": "cook_new",
      "title": "Cook Something New!",
      "subtitle": "Add another emoji to \(query) to make a new sticker",
      "icon": ["path": "cook-new.png"],
      "variables": ["emoji2": query]
    ]
    responseItems.insert(freshCook, at: 0)
  }
  if alfredItems.isEmpty {
    let sorry: [String: Any] = [
      "valid": false,
      "title": "Sorry, Empty Kitchen!",
      "icon": ["path": "empty-kitchen.png"]
    ]
    responseItems.insert(sorry, at: 0)
  }

  try updateFridge(
    kitchenDataDir: kitchenDataDir,
    newItems: alfredItems
  )
  let tmpPath = try writeTmpJson(
    kitchenDataDir: kitchenDataDir,
    items: responseItems
  )
  return tmpPath
}

do {
  let path = try runKitchen()
  print(path)
} catch {
  let kitchenDataDir = CommandLine.arguments.count >= 4
    ? CommandLine.arguments[3]
    : ""
  let message = (error as? KitchenError).map {
    if case let .message(msg) = $0 {
      return msg
    }
    return "Unknown error."
  } ?? error.localizedDescription

  if !kitchenDataDir.isEmpty {
    try? ensureDirectory(kitchenDataDir)
    _ = try? writeTmpJson(
      kitchenDataDir: kitchenDataDir,
      items: [["title": "Error: \(message)"]]
    )
  }
  fputs("Error: \(message)\n", stderr)
  exit(1)
}
